### ISSUE TITLE
logdet evals to -inf in an overly pessimistic way

### ISSUE THREAD

**Issue by JesseLivezey** [pytorch/pytorch#18450]
## 🐛 Bug

`logdet` returns `-inf` when it should be possible to calculate a well defined log-determinant. I haven't confirmed, but this is probably because this line
https://github.com/pytorch/pytorch/blob/f268370b42bc26d4e5d0fd88eb79b44ae162dec5/aten/src/ATen/native/LinearAlgebra.cpp#L65
is not stable and is used in the following if/else
https://github.com/pytorch/pytorch/blob/f268370b42bc26d4e5d0fd88eb79b44ae162dec5/aten/src/ATen/native/LinearAlgebra.cpp#L67

## To Reproduce

```python
import numpy as np
import torch

evals = torch.logspace(-4, -3, 100, dtype=torch.double)
cov = torch.diag(evals)

print(torch.logdet(cov))
print(torch.sum(torch.log(torch.svd(cov)[1])))
print(torch.log(torch.prod(torch.svd(cov)[1])))
print(np.linalg.slogdet(cov)[1])
```

## Expected behavior

`torch.logdet` should be more similar to sum(log()) than to log(prod()).

## Environment

PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Linux Mint 18.1 Serena
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: TITAN Xp

Nvidia driver version: 396.26
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6.0.21
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7.0.5

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.0.1.post2
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.10           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.0.1           py3.6_cuda9.0.176_cudnn7.4.2_2    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch

**Comment by SsnL** [comment]
lol we posted literally 10min apart. see https://github.com/pytorch/pytorch/issues/18448

and pr is https://github.com/pytorch/pytorch/pull/18449

**Comment by SsnL** [comment]
would you kindly comment on https://github.com/pytorch/pytorch/pull/18449 to see if it makes sense to you? I tested your repro case on my machine with the patch, and it seems to solve the issue.
```

In [15]: evals = torch.logspace(-4, -3, 100, dtype=torch.double)
    ...: cov = torch.diag(evals)
    ...:

In [16]:

In [16]: print(torch.logdet(cov))
    ...: print(torch.sum(torch.log(torch.svd(cov)[1])))
    ...: print(torch.log(torch.prod(torch.svd(cov)[1])))
tensor(-805.9048, dtype=torch.float64)
tensor(-805.9048, dtype=torch.float64)
tensor(-inf, dtype=torch.float64)

```

**Comment by SsnL** [comment]
btw, you can use `slogdet` as a workaround if you don't want to build from source

**Comment by JesseLivezey** [comment]
Thanks, looks good!

**Comment by JesseLivezey** [comment]
Closed by #18449
